pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,405 @@
+import functools
+from collections import defaultdict
+
+
+__all__ = (
+    "reduce_number_of_fmms",
+    )
+
+
+def reduce_number_of_fmms(int_gs, source_dependent_variables):
+    """
+    Reduce the number of FMMs needed for a system of
+    :class:`~pytential.symbolic.primitives.IntG` objects.
+
+    This is done by converting the ``IntG`` object to a matrix of polynomials
+    with d variables corresponding to d dimensions, where each variable represents
+    a (target) derivative operator along one of the axes. All the properties of
+    derivative operator that we want are reflected in the properties of the
+    polynomial including addition, multiplication and exact polynomial division.
+
+    This matrix is factored into two matrices, where the left hand side matrix
+    represents a transformation at the target, and the right hand side matrix
+    represents a transformation at the source.
+
+    If the expressions given are not linear, then the input expressions are
+    returned as is.
+
+    :arg int_gs: list of ``IntG`` objects.
+
+    :arg source_dependent_variables: list of :class:`pymbolic.primtives.Expression`
+        objects. When reducing FMMs, consider only these variables as dependent
+        on source. For eg: densities, source derivative vectors.
+
+    Note: there is no argument for target-dependent variables as the algorithm
+    assumes that there are no target-dependent-variables passed to this function.
+    (where a "target-dependent variable" represents a function discretized on the
+    targets)
+    """
+
+    dim = int_gs[0].target_kernel.dim
+    axis_vars = sympy.symbols(f"_x0:{dim}")
+
+    # A high level driver for this function should send int_gs that are common.
+    assert _check_int_gs_common(int_gs)
+
+    try:
+        mat, source_exprs = _create_matrix(int_gs, source_dependent_variables,
+            axis_vars)
+    except ValueError:
+        return int_gs
+
+    mat = sympy.Matrix(mat)
+    # Factor the matrix into two
+    try:
+        lhs_mat = _factor_left(mat, axis_vars)
+        rhs_mat = _factor_right(mat, lhs_mat)
+    except ValueError:
+        return int_gs
+
+    # If there are n inputs and m outputs,
+    #
+    #  - matrix = R^{m x n},
+    #  - LHS = R^{m x k},
+    #  - RHS = R^{k x n}.
+    #
+    # If k is not greater than or equal to n we are gaining nothing.
+    # Return as is.
+    if rhs_mat.shape[0] >= mat.shape[0]:
+        return int_gs
+
+    # Create SymPy Polynomial objects
+    rhs_mat = rhs_mat.applyfunc(lambda x: x.as_poly(*axis_vars, domain=sympy.EX))
+    lhs_mat = lhs_mat.applyfunc(lambda x: x.as_poly(*axis_vars, domain=sympy.EX))
+
+    base_kernel = int_gs[0].source_kernels[0].get_base_kernel()
+    base_int_g = int_gs[0].copy(target_kernel=base_kernel,
+            source_kernels=(base_kernel,), densities=(1,))
+
+    # Convert each element in the RHS matrix to IntGs
+    rhs_mat_int_gs = [[_convert_source_poly_to_int_g(poly, base_int_g, axis_vars)
```

Comment:
```suggestion
    rhs_mat_int_gs = [[_convert_source_poly_to_int_g_derivs(poly, base_int_g, axis_vars)
```

Reply:
Done in cf38f70832749e8cc94110a175e21cedde306992